In [39]:
# loading required libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [40]:
# loading pre-processed dataset
data = pd.read_csv("pre_processed_data.csv")
data.head()

area_code     area  year_code  year  item_code   item  producer_price  \
0          1  Armenia       1997  1997         15  Wheat           225.3   
1          1  Armenia       1998  1998         15  Wheat           205.6   
2          1  Armenia       1999  1999         15  Wheat           177.5   
3          1  Armenia       2000  2000         15  Wheat           163.5   
4          1  Armenia       2001  2001         15  Wheat           166.0   

   producer_price_index  area_harvested  production  ...  \
0                 89.40        108000.0    183700.0  ...   
1                 83.91        118300.0    244300.0  ...   
2                 76.79        110129.0    214380.0  ...   
3                 71.30        106581.0    177762.0  ...   
4                 74.51        108554.0    241679.0  ...   

   total_pesticide_use_per_value_of_agri_production  \
0                                              0.05   
1                                              0.06   
2                                              0.07   
3                                              0.08   
4                                              0.09   

   total_pesticide_export_quantity  total_pesticide_export_value  \
0                            3.604                         3.687   
1                            3.604                         3.687   
2                            3.604                         3.687   
3                            3.604                         3.687   
4                            1.145                         9.068   

   total_pesticide_import_quantity  total_pesticide_import_value  \
0                          117.861                       298.379   
1                          207.171                       340.124   
2                          296.481                       381.869   
3                          160.902                       482.955   
4                          200.884                       698.022   

   temp_change_dec_jan_feb  temp_change_mar_apr_may  temp_change_jun_jul_aug  \
0                    2.166                   -0.319                    0.870   
1                   -0.259                    1.483                    2.145   
2                    3.535                    0.814                    1.611   
3                    1.148                    0.723                    2.166   
4                    1.818                    2.072                    1.578   

   temp_change_sep_oct_nov  temp_change_meteorological_year  
0                    0.184                            0.725  
1                    2.027                            1.349  
2                    0.264                            1.556  
3                    0.364                            1.100  
4                    0.409                            1.469  

[5 rows x 97 columns]

In [41]:
data.shape

(211006, 97)

In [4]:
# Columns with > 10% missing values in descending order
prop_na = (data.isna().mean()*100).sort_values(ascending=False)
prop_na = prop_na.loc[prop_na > 10]
prop_na

govt_expenditure_on_ag_forest_fish             50.000000
producer_price                                 45.325251
afs_employment_share_in_total_employment       42.631489
credit_to_ag_forest_fish_share_total_credit    41.298352
credit_to_ag_forest_fish                       41.298352
aoi_credit_to_ag_forest_fish                   41.298352
total_employment_afs                           40.752396
agri_orientation_index_govt_expenditure        33.370615
total_govt_expenditure                         33.291944
area_temporary_crops                           31.785826
phosphorus_production                          28.609139
export_quantity                                27.330976
export_value                                   27.128612
gross_production_value                         26.565595
total_credit                                   26.085988
nitrogen_production                            26.032435
import_quantity                                25.894524
import_value                   

In [42]:
# list of feature names
feature_list = list(prop_na.keys())

# list of proportion of missing values
prop_list = list(round(val,3) for val in prop_na.values)

# Initialize lists to store the number of countries and items with non-missing data
num_countries = []
num_items = []

# Loop over each feature to calculate the number of unique countries and items
# for which the data is present (i.e., not NA)
for feature in feature_list:
    country_count = len(data['area'].loc[data[feature].notna()].unique())
    num_countries.append(country_count)
    item_count = len(data['item'].loc[data[feature].notna()].unique())
    num_items.append(item_count)

# Create a summary DataFrame with feature name, number of countries/items with data, 
# and % missing
df = pd.DataFrame({
    'Feature': feature_list,
    'Number_of_countries': num_countries,
    'Number_of_items': num_items,
    'Percent_NA': prop_list
})

df


Feature  Number_of_countries  \
0                                producer_price                  146   
1      afs_employment_share_in_total_employment                  123   
2   credit_to_ag_forest_fish_share_total_credit                  118   
3                      credit_to_ag_forest_fish                  118   
4                  aoi_credit_to_ag_forest_fish                  118   
5                          total_employment_afs                  129   
6       agri_orientation_index_govt_expenditure                  149   
7            govt_expenditure_on_ag_forest_fish                  142   
8                        total_govt_expenditure                  149   
9                          area_temporary_crops                  149   
10                        phosphorus_production                  149   
11                              export_quantity                  149   
12                                 export_value                  149   
13                       gross_production_value                  145   
14                                 total_credit                  118   
15                          nitrogen_production                  149   
16                              import_quantity                  149   
17                                 import_value                  149   
18                       gross_production_index                  149   
19                                        yield                  149   
20                               area_harvested                  149   

    Number_of_items  Percent_NA  
0               133      45.325  
1               134      42.631  
2               134      41.298  
3               134      41.298  
4               134      41.298  
5               134      40.752  
6               134      33.371  
7               134      33.360  
8               134      33.292  
9               134      31.786  
10              134      28.609  
11              132      27.331  
12              132      27.129  
13              129      26.566  
14              134      26.086  
15              134      26.032  
16              132      25.895  
17              132      25.794  
18              129      13.338  
19              130      11.239  
20              130      10.744

The five features with the highest proportion of missing data are:  
- govt_expenditure_on_ag_forest_fish, 
- producer_price, 
- credit_to_ag_forest_fish,
- afs_employment_share_in_total_employment, and
- total_employment_afs 

Among these, the missing values in the **govt_expenditure_on_ag_forest_fish** column 
can be partially addressed by leveraging the following related variables: 
**agri_orientation_index_govt_expenditure, ag_forest_fish_share_in_total_gdp, and total_govt_expenditure**.

The Agricultural Orientation Index (AOI) for governmental expenditure is defined as:
$$
AOI = \frac{\text{Agriculture's share of government expenditure}}{\text{Agriculture's share of GDP}}
$$

And, 
$$
\text{Agriculture's share of government expenditure} = \frac{\text{govt\_expenditure\_on\_ag\_forest\_fish}}{\text{total\_govt\_expenditure}}
$$

Hence, 
$$
\text{govt\_expenditure\_on\_ag\_forest\_fish} = \\
\text{(agri\_orientation\_index\_govt\_expenditure)} \times \text{(ag\_forest\_fish\_share\_in\_total\_gdp)} \times \text{(total\_govt\_expenditure)}
$$


In [43]:
# Estimate missing values in 'govt_expenditure_on_ag_forest_fish' 
# using the formula above
estimated_govt_expenditure = (
    data['agri_orientation_index_govt_expenditure'] * 
    data['ag_forest_fish_share_in_total_gdp'] * 
    data['total_govt_expenditure']
    )

# Fill missing values in 'govt_expenditure_on_ag_forest_fish' with the estimated values,
# but only where all three input components are non-missing (non-NaN)
data['govt_expenditure_on_ag_forest_fish'] = data['govt_expenditure_on_ag_forest_fish'].fillna(
    estimated_govt_expenditure.where(
        data[[
            'agri_orientation_index_govt_expenditure', 
            'ag_forest_fish_share_in_total_gdp', 
            'total_govt_expenditure'
            ]].notna().all(axis=1) # Ensures all required inputs are available
    ) 
)

# Recalculate missing value proportions for all columns
prop_na = (data.isna().mean()*100).sort_values(ascending=False)

# Keep only those features where more than 10% of the data is missing
prop_na = prop_na.loc[prop_na > 10]

# Convert column names and their missing value percentages to lists
feature_list = list(prop_na.keys())
prop_list = list(round(val,3) for val in prop_na.values)

# Count how many unique countries and items are present for each feature with >10% missing data
num_countries = []
num_items = []
for feature in feature_list:
    country_count = len(data['area'].loc[data[feature].notna()].unique())
    num_countries.append(country_count)
    item_count = len(data['item'].loc[data[feature].notna()].unique())
    num_items.append(item_count)

# Create a summary DataFrame showing missing data patterns
df = pd.DataFrame({
    'Feature': feature_list,
    'Number_of_countries': num_countries,
    'Number_of_items': num_items,
    'Percent_NA': prop_list
})

df

Feature  Number_of_countries  \
0                                producer_price                  146   
1      afs_employment_share_in_total_employment                  123   
2   credit_to_ag_forest_fish_share_total_credit                  118   
3                      credit_to_ag_forest_fish                  118   
4                  aoi_credit_to_ag_forest_fish                  118   
5                          total_employment_afs                  129   
6       agri_orientation_index_govt_expenditure                  149   
7            govt_expenditure_on_ag_forest_fish                  149   
8                        total_govt_expenditure                  149   
9                          area_temporary_crops                  149   
10                        phosphorus_production                  149   
11                              export_quantity                  149   
12                                 export_value                  149   
13                       gross_production_value                  145   
14                                 total_credit                  118   
15                          nitrogen_production                  149   
16                              import_quantity                  149   
17                                 import_value                  149   
18                       gross_production_index                  149   
19                                        yield                  149   
20                               area_harvested                  149   

    Number_of_items  Percent_NA  
0               133      45.325  
1               134      42.631  
2               134      41.298  
3               134      41.298  
4               134      41.298  
5               134      40.752  
6               134      33.371  
7               134      33.360  
8               134      33.292  
9               134      31.786  
10              134      28.609  
11              132      27.331  
12              132      27.129  
13              129      26.566  
14              134      26.086  
15              134      26.032  
16              132      25.895  
17              132      25.794  
18              129      13.338  
19              130      11.239  
20              130      10.744

So, we were able to reduce the amount of missing data in feature 'govt_expenditure_on_ag_forest_fish' 
from 50.00% to 33.36%. 

Next, the other columns with high missing data have data for 118, 123, 129 countries 
out of 149 total. In order to reduce the amount of missingness in these columns, 
we can try to find the common countries that have missing data for all these columns.

In [45]:
# Get list of countries (areas) with non-missing values in each respective column
area_afs = list(data.loc[data['total_employment_afs'].notna()]['area'].unique())
area_credit = list(data.loc[data['credit_to_ag_forest_fish'].notna()]['area'].unique())
area_afs_share = list(data.loc[data['afs_employment_share_in_total_employment'].notna()]['area'].unique())

# All countries present in the dataset
area_total = list(data['area'].unique())

# Identify countries missing each variable by subtracting from the full country list
missing_credit = set(area_total) - set(area_credit)
missing_afs = set(area_total) - set(area_afs)
missing_afs_share = set(area_total) - set(area_afs_share)

# Find countries missing both credit and FDI data
common_area = set(missing_credit).intersection(missing_afs)
print(f"Number of common countries that are missing in both 'total_employment_afs' and 'credit_to_ag_forest_fish_2015_usd': {len(common_area)}")

# Find countries missing both credit and AFS employment share data
common_area = set(missing_credit).intersection(missing_afs_share)
print(f"Number of common countries that are missing in both 'afs_employment_share_in_total_employment' and 'credit_to_ag_forest_fish_2015_usd': {len(common_area)}")

# Find countries missing both FDI and AFS employment share data
common_area = set(missing_afs_share).intersection(missing_afs)
print(f"Number of common countries that are missing in both 'total_employment_afs' and 'afs_employment_share_in_total_employment': {len(common_area)}")

Number of common countries that are missing in both 'total_employment_afs' and 'credit_to_ag_forest_fish_2015_usd': 0
Number of common countries that are missing in both 'afs_employment_share_in_total_employment' and 'credit_to_ag_forest_fish_2015_usd': 0
Number of common countries that are missing in both 'total_employment_afs' and 'afs_employment_share_in_total_employment': 20


Unfortunately, there is limited overlap among these columns in terms of countries 
with missing data.

**Furthermore, due to the high sparsity in the producer_price column, we chose to** 
**remove it as well. Instead, we plan to forecast the producer_price_index, for** 
**which data coverage is more consistent across countries and years.** 

In [ ]:
# Dropping producer_price column
data_1 = data.drop('producer_price', axis = 1)
data_1.shape

(211006, 96)

Due to substantial missingness, we dropped the FDI-related columns from the dataset. 
However, we introduced a new feature, value_added_aff_per_total_fdi, calculated 
as the ratio of value_added_ag_forest_fish to total_fdi_inflows for each country-year 
pair. These two variables have relatively high data availability, making the 
derived ratio more reliable. This new feature provides an indication of how the 
agriculture sector contributes to the national economy in relation to varying 
levels of FDI inflows.

In [48]:
# To avoid division-by-zero errors, we only compute the ratio when 'total_fdi_inflows' is non-zero.
# If 'total_fdi_inflows' is zero, the result is set to NaN.
data_1['value_added_aff_per_total_fdi'] = np.where(
    data_1['total_fdi_inflows'] != 0,
    data_1['value_added_ag_forest_fish'] / data_1['total_fdi_inflows'],
    np.nan
)
data_1.shape

(211006, 97)

In [49]:
# Recalculate missing value proportions for all columns
prop_na_1 = (data_1.isna().mean()*100).sort_values(ascending=False)

# Keep only those features where more than 10% of the data is missing
prop_na_1 = prop_na_1.loc[prop_na_1 > 10]

# Convert column names and their missing value percentages to lists
feature_list_1 = list(prop_na_1.keys())
prop_list_1 = list(round(val,3) for val in prop_na_1.values)

# Count how many unique countries and items are present for each feature with >10% missing data
num_countries = []
num_items = []
for feature in feature_list_1:
    country_count = len(data_1['area'].loc[data_1[feature].notna()].unique())
    num_countries.append(country_count)
    item_count = len(data_1['item'].loc[data_1[feature].notna()].unique())
    num_items.append(item_count)

# Create a summary DataFrame showing missing data patterns
df_1 = pd.DataFrame({
    'Feature': feature_list_1,
    'Number_of_countries': num_countries,
    'Number_of_items': num_items,
    'Percent_NA': prop_list_1
})

df_1

Feature  Number_of_countries  \
0      afs_employment_share_in_total_employment                  123   
1                  aoi_credit_to_ag_forest_fish                  118   
2                      credit_to_ag_forest_fish                  118   
3   credit_to_ag_forest_fish_share_total_credit                  118   
4                          total_employment_afs                  129   
5       agri_orientation_index_govt_expenditure                  149   
6            govt_expenditure_on_ag_forest_fish                  149   
7                        total_govt_expenditure                  149   
8                          area_temporary_crops                  149   
9                         phosphorus_production                  149   
10                              export_quantity                  149   
11                                 export_value                  149   
12                       gross_production_value                  145   
13                                 total_credit                  118   
14                          nitrogen_production                  149   
15                              import_quantity                  149   
16                                 import_value                  149   
17                       gross_production_index                  149   
18                                        yield                  149   
19                               area_harvested                  149   

    Number_of_items  Percent_NA  
0               134      42.631  
1               134      41.298  
2               134      41.298  
3               134      41.298  
4               134      40.752  
5               134      33.371  
6               134      33.360  
7               134      33.292  
8               134      31.786  
9               134      28.609  
10              132      27.331  
11              132      27.129  
12              129      26.566  
13              134      26.086  
14              134      26.032  
15              132      25.895  
16              132      25.794  
17              129      13.338  
18              130      11.239  
19              130      10.744

At this stage, since no further augmentation of the missing data is possible 
through alternative sources or transformations, we must now proceed with data 
imputation techniques to address the remaining gaps. Prior to implementing these 
methods, it is essential to incorporate indicator columns into the dataset. These 
columns categorize countries based on geographic (e.g., region, sub-region) and 
economic characteristics (e.g., least_developed_country, european_union_country, 
low_income_food_deficit_country, etc.).

These groupings are crucial for informed imputation, as they allow us to impute 
missing values using data from comparable countries within the same group. For 
instance, it would be inappropriate to estimate missing values for an African 
country using data from European countries, given the distinct economic and regional 
contexts. Group-based imputation ensures that the assumptions underlying the 
process remain valid and contextually grounded.

In [ ]:
# loading required package
from sqlalchemy import create_engine

In [52]:
# creating a connection to the database
engine = create_engine('postgresql+psycopg2://postgres:1995@localhost:5432/faostat_ms_dsci_project')

In [ ]:
# importing data 
indicator_table = pd.read_sql_table('indicator_columns', engine)
indicator_table.head()

area_code_m49                 area    region       sub_region  \
0              4          Afghanistan      Asia       South Asia   
1              8              Albania    Europe  Southern Europe   
2             12              Algeria    Africa  Northern Africa   
3             24               Angola    Africa    Middle Africa   
4             28  Antigua and Barbuda  Americas        Carribean   

   european_union_country  least_developed_country  \
0                     0.0                      1.0   
1                     0.0                      0.0   
2                     0.0                      0.0   
3                     0.0                      1.0   
4                     0.0                      0.0   

   land_locked_developing_country  small_island_developing_state  \
0                             1.0                            0.0   
1                             0.0                            0.0   
2                             0.0                            0.0   
3                             0.0                            0.0   
4                             0.0                            1.0   

   low_income_food_deficit_country  net_food_importing_developing_country  
0                              1.0                                    1.0  
1                              0.0                                    0.0  
2                              0.0                                    0.0  
3                              0.0                                    1.0  
4                              0.0                                    1.0

In [ ]:
# adding indicator columns to our dataseet
merged_data = pd.merge(
    data_1, indicator_table,
    on='area',
    how='left'
)
merged_data.head()

area_code     area  year_code  year  item_code   item  \
0          1  Armenia       1997  1997         15  Wheat   
1          1  Armenia       1998  1998         15  Wheat   
2          1  Armenia       1999  1999         15  Wheat   
3          1  Armenia       2000  2000         15  Wheat   
4          1  Armenia       2001  2001         15  Wheat   

   producer_price_index  area_harvested  production   yield  ...  \
0                 89.40        108000.0    183700.0  1700.9  ...   
1                 83.91        118300.0    244300.0  2065.1  ...   
2                 76.79        110129.0    214380.0  1946.6  ...   
3                 71.30        106581.0    177762.0  1667.9  ...   
4                 74.51        108554.0    241679.0  2226.3  ...   

   value_added_aff_per_total_fdi  area_code_m49  region    sub_region  \
0                       6.445164             51    Asia  Western Asia   
1                       1.749632             51    Asia  Western Asia   
2                       3.192425             51    Asia  Western Asia   
3                       3.616866             51    Asia  Western Asia   
4                       6.091945             51    Asia  Western Asia   

   european_union_country  least_developed_country  \
0                     0.0                      0.0   
1                     0.0                      0.0   
2                     0.0                      0.0   
3                     0.0                      0.0   
4                     0.0                      0.0   

   land_locked_developing_country  small_island_developing_state  \
0                             1.0                            0.0   
1                             1.0                            0.0   
2                             1.0                            0.0   
3                             1.0                            0.0   
4                             1.0                            0.0   

   low_income_food_deficit_country  net_food_importing_developing_country  
0                              0.0                                    0.0  
1                              0.0                                    0.0  
2                              0.0                                    0.0  
3                              0.0                                    0.0  
4                              0.0                                    0.0  

[5 rows x 106 columns]

In [55]:
merged_data['region'].unique()

array(['Asia', 'Europe', 'Africa', 'Americas', 'Oceania'], dtype=object)

In [56]:
len(set(data_1['area']).intersection(set(indicator_table['area'])))

149

In [57]:
set(data_1['area']) - set(indicator_table['area'])

set()